File: backend/models/protocol.py
from typing import Literal

from pydantic import BaseModel

from backend.models.action import ActionIntent
from backend.models.dispute import DisputeReason
from backend.models.event import Event
from backend.models.seat import AckPosture


class WSInbound(BaseModel):
    msg_type: Literal["action", "ack", "nack", "dispute", "chat",
                       "set_ack_posture", "ping"]
    intent: ActionIntent | None = None
    action_id: str | None = None
    reason: DisputeReason | None = None
    reason_text: str | None = None
    text: str | None = None
    channel: str | None = None
    ack_posture: AckPosture | None = None


class WSOutbound(BaseModel):
    msg_type: Literal["event", "state_sync", "error", "pong"]
    event: Event | None = None
    state: dict | None = None
    error: str | None = None
    error_code: str | None = None

File: frontend/src/ws/protocol.ts
import type {
  ActionIntent,
  AckPosture,
  Event,
  TableState,
} from "../types/models";
import type { DisputeReason } from "../types/enums";

export type WSInbound =
  | { msg_type: "action"; intent: ActionIntent }
  | { msg_type: "ack"; action_id: string }
  | {
      msg_type: "nack";
      action_id: string;
      reason?: DisputeReason;
      reason_text?: string;
    }
  | {
      msg_type: "dispute";
      action_id: string;
      reason?: DisputeReason;
      reason_text?: string;
    }
  | { msg_type: "chat"; text: string; channel?: string }
  | { msg_type: "set_ack_posture"; ack_posture: AckPosture }
  | { msg_type: "ping" };

export type WSOutbound =
  | { msg_type: "state_sync"; state: TableState }
  | { msg_type: "event"; event: Event }
  | { msg_type: "error"; error: string; error_code: string }
  | { msg_type: "pong" };
